stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/list/adapter/ChannelListItemAdapter.kt

```diff
@@ -12,32 +11,55 @@ import io.getstream.chat.android.ui.channel.list.adapter.viewholder.ChannelListL
 import io.getstream.chat.android.ui.utils.extensions.cast
 import io.getstream.chat.android.ui.utils.extensions.diff
 import io.getstream.chat.android.ui.utils.extensions.firstOrDefault
+import io.getstream.chat.android.ui.utils.extensions.safeCast
 
 internal class ChannelListItemAdapter(var style: ChannelListViewStyle) :
-    ListAdapter<Channel, BaseChannelListItemViewHolder>(DIFF_CALLBACK) {
+    ListAdapter<ChannelListItem, BaseChannelListItemViewHolder>(DIFF_CALLBACK) {
 
     var viewHolderFactory: ChannelListItemViewHolderFactory = ChannelListItemViewHolderFactory()
 
     val listenerContainer: ChannelListListenerContainer = ChannelListListenerContainer()
 
-    var endReached: Boolean = true
-        set(value) {
-            if (value) {
-                // if we've reached the end, remove the last item
-                notifyItemRemoved(itemCount)
-            }
-            field = value
-        }
-
     companion object {
-        val DIFF_CALLBACK: DiffUtil.ItemCallback<Channel> = object : DiffUtil.ItemCallback<Channel>() {
-            override fun areItemsTheSame(oldItem: Channel, newItem: Channel): Boolean = oldItem.cid == newItem.cid
-
-            override fun areContentsTheSame(oldItem: Channel, newItem: Channel): Boolean =
-                !oldItem.diff(newItem).hasDifference()
-
-            override fun getChangePayload(oldItem: Channel, newItem: Channel): Any = oldItem.diff(newItem)
-        }
+        val DIFF_CALLBACK: DiffUtil.ItemCallback<ChannelListItem> =
+            object : DiffUtil.ItemCallback<ChannelListItem>() {
+                override fun areItemsTheSame(oldItem: ChannelListItem, newItem: ChannelListItem): Boolean {
+                    if (oldItem::class != newItem::class) {
```

Comment:
You can use only boolean logic here:
```
return oldItem::class == newItem::class && (oldItem is ChannelListItem.ChannelItem && oldItem.channel.cid == newItem.safeCast<ChannelListItem.ChannelItem>()?.channel?.cid || oldItem !is ChannelListItem.ChannelItem)
```

Reply:
It is a good idea to extract some parts of the boolean logic to a method to avoid a super big line.

Reply:
I like the structure in this `ItemCallback` implementation as-is, it handles cases clearly, and it will be easy to update clearly if we ever need to add a new item type.

Reply:
Just going for some readability & taking advantage of smart casting from `when`. I want to avoid the big line, but also avoid the manual cast of `oldItem` and evaluation of `oldItem.channel.cid == newItem.safeCast<ChannelListItem.ChannelItem>()?.channel?.cid`, which would happen if I extracted both into named variables (`return sameClass && sameCid`).